### ISSUE TITLE
dependency('', fallback: ['something', 'dep']) doesn't look for the fallback

### ISSUE THREAD

**Issue by ystreet** [mesonbuild/meson#3754]
As it says on the tin nothing is output about the dependency or looking for the fallback subproject if the '' dependency is searched for.

Changing the dependency name to some not-found dependency results in:
```
Dependency no-find found: NO
Also couldn't find a fallback subproject in subprojects/something for the dependency no-find 
Reason: Subproject directory 'subprojects/something' does not exist and cannot be downloaded:
No something.wrap found for 'subprojects/something'
```

**Comment by jon-turney** [comment]
This is interesting.  `dependency('')` is handled specially since #2615 to avoid reporting searching for `''`, but `fallback:` should be taken into consideration.

Notwithstanding what the documentation for `subproject()` currently says, it might well be clearer to write `subproject('subproject').get_variable('dep')` rather than `dependency('', fallback: ['subproject', 'dep'])` when the dependency does not have a name under which it might be found as an installed, external dependency, and only exists as a subproject.

**Comment by nirbheek** [comment]
The reason why a dependency fallback is useful over an explicit subproject call is for two reasons:

1. When the subproject is optional, and you want to handle that gracefully within the build file
2. When you want to use wrap mode `nodownload`, because it can only work with dependency fallbacks

**Comment by jon-turney** [comment]
>   When the subproject is optional, and you want to handle that gracefully within the build file

Ah, good point.  I updated the PR to drop the documentation change and added some test coverage of that.

On second thoughts, I'm wondering if we shouldn't force this kind of dependency to be given a name though, since all the associated logging wants to use it.

**Comment by jon-turney** [comment]
Closed by #3778
